src/main/java/com/google/devtools/build/lib/remote/ByteStreamBuildEventArtifactUploader.java

```diff
@@ -201,6 +199,29 @@ public DigestFunction.Value getDigestFunction() {
   private PathMetadata readPathMetadata(Path path, LocalFile file) throws IOException {
     DigestUtil digestUtil = new DigestUtil(xattrProvider, path.getFileSystem().getDigestFunction());
 
+    if (remoteBuildEventUploadMode == RemoteBuildEventUploadMode.MINIMAL) {
+      // Omit all unimportant paths
+      if (!(
+        // Logs (regex needed for test logs as they lack a dedicated `LocalFileType`)
+        TEST_LOG_PATTERN.matcher(path.getPathString()).matches()
+        || file.type == LocalFileType.LOG
+        || file.type == LocalFileType.PERFORMANCE_LOG
+        // std(out|err)
+        || file.type == LocalFileType.STDOUT
+        || file.type == LocalFileType.STDERR
+      )) {
+        return new PathMetadata(
+            path,
+            digestUtil.compute(path),
+            /* directory= */ false,
+            /* symlink= */ false,
+            /* remote= */ false,
+            /* omitted= */ true,
+            /* isBuildToolLog= */ false,
+            digestUtil.getDigestFunction());
+      }
+    }
```

Comment:
I find myself asking the same question. This went through a number of iterations, as I discovered more and more requirements that incrementally eliminated any kind of potential optimisation.

I'm going to close this. There is room for improvement in the bytestream uploader (I had a stab at some in #23252 but hit a NPE I could not track down, at face value the area seems like it would benefit from a refactor and perhaps alignment across full and minimal upload modes).

Reply:
FWIW, I don't think the uploader should be digesting files at all; all output files have already been digested (and the digests cached in memory as a `FileArtifactValue`) during action execution, so we should figure out a way to plumb that data into the uploader.